src/bootstrap/builder.rs

```diff
@@ -360,10 +367,19 @@ impl StepDescription {
         for (desc, should_run) in v.iter().zip(&should_runs) {
             let pathsets = should_run.pathset_for_paths_removing_matches(&mut paths, desc.kind);
             if !pathsets.is_empty() {
-                desc.maybe_run(builder, pathsets);
+                hosts_check.push(desc.maybe_run(builder, pathsets));
             }
         }
 
+        // sanity checks on hosts
+        if !hosts_check.contains(&Some(())) {
+            eprintln!(
+                "`x.py {}` run with empty `host` parameter. Either set it or leave it out for default value.",
```

Comment:
This error message seems to believe that this can only happen if we have hosts as an empty parameter, but I'm not convinced by skimming the code above that's the only way to hit this. Can we document the rationale for that conclusion in a comment, plus why we don't validate host='' at flag parsing time?

Reply:
As far as I understand, there are three cases:
1) `./x.py` is invoked with no `host` parameter. This will make the host value default to this machine.
2) `./x.py` is invoked with an explicitely set empty `host` parameter, **but also** an explicitely set non-empty `target` parameter. This will set the host value to the target value.
3) `./x.py` is invoked with an explicitely set empty `host` parameter and no `target` parameter.

Case 3 is what we are fixing here. I think I achieved this by the check on [line 294](https://github.com/rust-lang/rust/blob/0118a8b34580ed7b34d7f8071a6a78b97f38ab65/src/bootstrap/builder.rs#L294). This also takes case 1 and 2 into consideration and does not treat those as bad cases.

> plus why we don't validate host='' at flag parsing time?

When is that - or where is that? :)

Reply:
Flag parsing is in flags.rs and config.rs, I believe. (Maybe we can put this check there, but I suspect that it's not that simple - it is sometimes valid to pass --host with an empty string. Not 100% confident about that though.

Reply:
Would you mind us keeping this check here ? This fixes the issue at hand and is tied together with the other logic of testing for empty host and target parameters.

Reply:
I am still not following the logic here, or exactly why you think this code specifically targets the case you mention. But the code here (https://github.com/rust-lang/rust/blob/master/src/bootstrap/config.rs#L1005) seems like it's able to directly detect that we're setting host + target to empty sets and error based on that.

Reply:
If you would rather have it there, I can try to look into that. It was just mentioned in the issue ( https://github.com/rust-lang/rust/issues/77906 ) that this would be the spot to fix it. And now it is fixed there.

Reply:
I'm not sure why @jyn514 suggested fixing #77906 here. It seems to me that *if* a fix is necessary at all, we would want to fix this in a more general manner -- for example, if nothing is run, we error rather than saying "all is successful", without trying to determine *why* nothing was run (e.g., exclude rules excluding all targets, etc.)

Reply:
> I'm not sure why @jyn514 suggested fixing https://github.com/rust-lang/rust/issues/77906 here. It seems to me that if a fix is necessary at all, we would want to fix this in a more general manner -- for example, if nothing is run, we error rather than saying "all is successful", without trying to determine why nothing was run (e.g., exclude rules excluding all targets, etc.)

I don't know why this place in the code was picked, but to me it seems like a fine place. We can check what we need to check, emit errors with a return code 1 and it fits some of the other checks that is being done here (and could be done). But how do we determine this? This is my second pull request for this project, so I don't imagine my opinion matters too much.

> for example, if nothing is run, we error rather than saying "all is successful", without trying to determine why nothing was run (e.g., exclude rules excluding all targets, etc.)

I am not sure I understand this. If nothing is run due to empty host or exclude rules or something third, I think it is important to tell the user that the build was unsuccessful and tell them why. The why is important because that gives them a path forward - a way to fix the issue. Do we disagree on this?

Reply:
> I am not sure I understand this. If nothing is run due to empty host or exclude rules or something third, I think it is important to tell the user that the build was unsuccessful and tell them why. The why is important because that gives them a path forward - a way to fix the issue. Do we disagree on this?

I don't think we agree that providing arguments that lead to a no-op build necessarily should error, but it doesn't seem bad to error in that case. Regardless, this PR is giving a very specific error message about providing empty hosts but isn't touching the code in a way that I can quickly say "that is the only way to reach this error message". Telling the user that they have empty hosts *when that may not be the case* is very confusing, and I cannot easily follow the conditions to in fact confirm that is the case for this error.